### PR TITLE
🤓 Record dependencies and result names in the definition after flow validation

### PR DESCRIPTION
**Pull request by rowanseymour**

### CODE REVIEW

File: extensions/transferto/action.go [hunk 1]
```diff
@@ -45,8 +45,8 @@ func NewTransferAirtimeAction(uuid flows.ActionUUID, amounts map[string]decimal.
 	}
 }
 
-// Validate validates our action is valid and has all the assets it needs
-func (a *TransferAirtimeAction) Validate(assets flows.SessionAssets, context *flows.ValidationContext) error {
+// Validate validates our action is valid
+func (a *TransferAirtimeAction) Validate() error {
```
Comment (rowanseymour) [inline extensions/transferto/action.go]: actions now only do structural validation on themselves - they're not responsible for determining if their dependencies exist - we do that centrally

File: flows/actions/testdata/add_contact_groups.json [hunk 2]
```diff
@@ -11,7 +11,7 @@
                 }
             ]
         },
-        "validation_error": "no such group with UUID '33382939-babf-4982-9395-8793feb4e7c6'"
+        "validation_error": "missing dependencies: group[uuid=33382939-babf-4982-9395-8793feb4e7c6,name=Climbers]"
```
Comment (rowanseymour) [inline flows/actions/testdata/add_contact_groups.json]: this should never end up being a user facing error.. assuming we track dependencies correctly

File: flows/definition/dependencies.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+			missing(ref)
+		} else {
+			d.Channels[i] = a.Reference()
+		}
+	}
+	for i, ref := range d.Fields {
+		a, err := sa.Fields().Get(ref.Key)
+
+		if err != nil {
+			// TODO for now if a field reference came from an expression (i.e. no name), we don't blow up if it's missing
```
Comment (rowanseymour) [inline flows/definition/dependencies.go]: chatted with @nicpottier and we can look at addressing this in the old world but there are just too many flows out there that likely to have expressions referencing fields that don't exist

File: flows/definition/flow.go [hunk 4]
```diff
@@ -95,12 +105,38 @@ func (f *flow) Validate(assets flows.SessionAssets, context *flows.ValidationCon
...
+		depStrings := make([]string, len(missingAssets))
+		for d := range missingAssets {
+			depStrings[d] = missingAssets[d].String()
+		}
+		return errors.Errorf("missing dependencies: %s", strings.Join(depStrings, ","))
+	}
+
 	f.valid = true
+	f.dependencies = deps
+	f.resultNames = f.ExtractResultNames()
```
Comment (rowanseymour) [inline flows/definition/flow.go]: I think this is threadsafe... it's possible that multiple threads validate this flow object at the same time, but it shouldn't matter

File: flows/definition/flow.go [hunk 5]
```diff
@@ -173,10 +209,12 @@ func (f *flow) ExtractDependencies() []assets.Reference {
 	dependencies := make([]assets.Reference, 0)
 	dependenciesSeen := make(map[string]bool)
 	addDependency := func(r assets.Reference) {
-		key := fmt.Sprintf("%s:%s", r.Type(), r.Identity())
-		if r != nil && !r.Variable() && !dependenciesSeen[key] {
-			dependencies = append(dependencies, r)
-			dependenciesSeen[key] = true
+		if !utils.IsNil(r) && !r.Variable() {
```
Comment (rowanseymour) [inline flows/definition/flow.go]: See https://github.com/nyaruka/goflow/issues/534

File: flows/actions/start_session.go [hunk 6]
```diff
@@ -53,15 +53,9 @@ func NewStartSessionAction(uuid flows.ActionUUID, flow *assets.FlowReference, ur
...
-// Validate validates our action is valid and has all the assets it needs
-func (a *StartSessionAction) Validate(assets flows.SessionAssets, context *flows.ValidationContext) error {
-	// check the flow exists and that it's valid
-	if err := a.validateFlow(assets, a.Flow, context); err != nil {
-		return err
-	}
-
-	// finally check that all the groups exist
-	return a.validateGroups(assets, a.Groups)
+// Validate validates our action is valid
```
Comment (nicpottier) [inline flows/actions/start_session.go]: Maybe no-op Validate should be part of BaseAction?

**Pull request merged**